sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamIOSourceRel.java

```diff
@@ -29,11 +30,17 @@
 /** BeamRelNode to replace a {@code TableScan} node. */
 public class BeamIOSourceRel extends TableScan implements BeamRelNode {
 
-  private BeamSqlTable sqlTable;
+  private final BeamSqlTable sqlTable;
+  private final Map<String, String> pipelineOptions;
```

Comment:
I would suggest to have `pipelineOptions` on system/session level. Think about that the query has multiple tables, for join, what if they have different settings?

Reply:
What do you mean by having `pipelineOptions` on the system/session level? These concepts aren't really different in Beam, there isn't a persistent system that can be configured separately from the session.

You can have multiple different `pipelineOptions` for tables if you have multiple schema. However that isn't supported and you'll find that the default `getPipelineOptions` asserts that every table has the same `pipelineOptions` map. As long as your tables are all on the same schema, they will have the same `pipelineOptions`.

If your question is why they are plumbed through here, that is because we don't have a single system level config that can be accessed from the BeamEnumerableConverter in every possible environment.

Reply:
There was a discussion before to expose `BeamSqlEnv` in `BeamRelNode` or `BeamRelNode.toPTransform()`. This may be an option to have SET in session level.

Reply:
I remember a discussion about removing `BeamSqlEnv` and replacing it with `meta/store` and `CalciteSchema` (which is implemented by `BeamCalciteSchema`), is that the one? I'm still not following how the concepts of system and session apply here.

Reply:
That's the thread. My though of `pipelineOptions on session` level is, only keep `pipelineOptions` in `BeamCalciteSchema`, and expose `BeamCalciteSchema` where is needed. 
I use `BeamSqlCli` on version 2.4.0, it refers to [these lines](https://github.com/apilloud/beam/blob/9c1cd62cdf568cdd3ebb87f44771ded298984fde/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/BeamSqlCli.java#L58-L63).

Reply:
That is basically what is done here. I think you are wondering how to get it into `BeamSqlCli`? I'm out of the office today (and possibly tomorrow) but I'll get that added to this PR.

I think you might like the JDBC interface we've been working on. It can run queries like `BeamSqlCli` but can also give you output, for example row count on INSERT INTO or actual rows on SELECT queries.